File: C#/WindowsPhone/LGame-XNA-lib/Loon.Action.Sprite.Node/LNAction.cs
using Loon.Utils;
namespace Loon.Action.Sprite.Node {
	

	public abstract class LNAction {
	
		public LNAction() {
			this._firstTick = true;
			this._isPause = false;
		}
	
		protected internal Easing _easing;
	
		protected internal float _duration;
	
		protected internal float _elapsed;
	
		protected internal bool _firstTick;
	
		protected internal bool _isEnd;
	
		protected internal bool _isPause;
	
		protected internal LNNode _target;
	
		public void AssignTarget(LNNode node) {
			this._target = node;
		}
	
		public void Pause() {
			this._isPause = true;
		}
	
		public void Resume() {
			this._isPause = false;
		}
	
		public virtual void SetTarget(LNNode node) {
			this._firstTick = true;
			this._isEnd = false;
			this._target = node;
		}
	
		public abstract LNAction Copy();
	
		public void Start() {
			this.SetTarget(this._target);
		}
	
		public virtual void Step(float dt) {
			if (!this._isPause) {
				if (this._firstTick) {
					this._firstTick = false;
					this._elapsed = 0f;
				} else {
					this._elapsed += dt;
				}
				float fx = 0;
				if (_easing != null) {
					fx = _easing.Ease((this._elapsed / this._duration), 1f);
				} else {
					fx = MathUtils.Min((this._elapsed / this._duration), 1f);
				}
				this.Update(fx);
			}
		}
	
		public void Reset() {
			this._elapsed = 0;
			this._isEnd = false;
		}
	
		public void Stop() {
			this._isEnd = true;
		}
	
		public virtual void Update(float time) {
		}
	
		public float GetDuration() {
			return this._duration;
		}
	
		public float GetElapsed() {
			return this._elapsed;
		}
	
		public bool IsEnd() {
			return this._isEnd;
		}
	
		public Easing GetEasing() {
			return _easing;
		}
	
		public void SetEasing(Easing e) {
			this._easing = e;
		}
	}
}

File: Java/OpenGL-1.0(old_ver)/Loon-backend-Android/src/loon/action/sprite/node/LNAction.java
/**
 * Copyright 2008 - 2012
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.3.3
 */
package loon.action.sprite.node;

import loon.utils.MathUtils;

public abstract class LNAction {

	protected Easing _easing;

	protected float _duration;

	protected float _elapsed;

	protected boolean _firstTick = true;

	protected boolean _isEnd;

	protected boolean _isPause = false;

	protected LNNode _target;

	public final void assignTarget(LNNode node) {
		this._target = node;
	}

	public void pause() {
		this._isPause = true;
	}

	public void resume() {
		this._isPause = false;
	}

	public void setTarget(LNNode node) {
		this._firstTick = true;
		this._isEnd = false;
		this._target = node;
	}

	public abstract LNAction copy();

	public void start() {
		this.setTarget(this._target);
	}

	public void step(float dt) {
		if (!this._isPause) {
			if (this._firstTick) {
				this._firstTick = false;
				this._elapsed = 0f;
			} else {
				this._elapsed += dt;
			}
			float fx = 0;
			if (_easing != null) {
				fx = _easing.ease((this._elapsed / this._duration), 1f);
			} else {
				fx = MathUtils.min((this._elapsed / this._duration), 1f);
			}
			this.update(fx);
		}
	}

	public void reset() {
		this._elapsed = 0;
		this._isEnd = false;
	}

	public void stop() {
		this._isEnd = true;
	}

	public void update(float time) {
	}

	public final float getDuration() {
		return this._duration;
	}

	public final float getElapsed() {
		return this._elapsed;
	}

	public final boolean isEnd() {
		return this._isEnd;
	}

	public Easing getEasing() {
		return _easing;
	}

	public void setEasing(Easing e) {
		this._easing = e;
	}
}
